### PR TITLE
datatable editing

### PR DESCRIPTION
**Pull request by rfontanarosa**
https://user-images.githubusercontent.com/10047945/157230762-434b5d03-4e6e-49e3-b70c-c1b867e3c449.mov

### CODE REVIEW

File: client/store/pages/originalDataPoint/actions/updateOriginalDataPoint.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ export const updateOriginalDataPoint = createAsyncThunk<
   OriginalDataPoint,
   { assessmentName: string; cycleName: string; originalDataPoint: OriginalDataPoint }
 >('originalDataPoint/put/byId', async ({ assessmentName, cycleName, originalDataPoint }) => {
-  const { data } = await axios.put(
+  const { data } = await axios.patch(
```
Comment (rfontanarosa) [inline client/store/pages/originalDataPoint/actions/updateOriginalDataPoint.ts:11]: use conventions

File: common/api/endpoint/ApiEndPoint.ts [hunk 2]
```diff
@@ -5,6 +5,11 @@ export const ApiEndPoint = {
   Init: {
     one: () => apiPath('init'),
   },
+  CycleData: {
+    PersistNode: {
+      one: () => apiPath('cycleData', 'persistNode'),
+    },
+  },
```
Comment (rfontanarosa) [inline common/api/endpoint/ApiEndPoint.ts:12]: how we should define the query params used for this (e.g. countryIso, assessmentName, cycleName...)

Comment (rfontanarosa) [reply common/api/endpoint/ApiEndPoint.ts:12]: @minotogna

**Comment by sorja** [comment]
lock table editing if isDataTableUpdating

File: client/pages/AssessmentSection/DataTable/Table/Row/RowData/Cell/useOnChange.ts [hunk 3]
```diff
@@ -1,92 +1,108 @@
...
+      })
+    )
+  }
+
+  const onChange: ChangeEventHandler<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement> = (event) => {
+    const { value } = event.target
+    _persistSanitizedValue(value)
+  }
+
+  // const onPaste: ClipboardEventHandler<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement> = (event) => {
```
Comment (rfontanarosa) [inline client/pages/AssessmentSection/DataTable/Table/Row/RowData/Cell/useOnChange.ts:63]: next pr implements onpaste

File: client/pages/AssessmentSection/DataTable/DataTable.tsx [hunk 4]
```diff
@@ -62,6 +64,9 @@ const DataTable: React.FC<Props> = (props) => {
     }
   }, [sectionName])
 
+  // Refetch on update
+  // TODO: Use websocket for this later
```
Comment (minotogna) [inline client/pages/AssessmentSection/DataTable/DataTable.tsx:68]: Remove this TODO

File: client/store/pages/assessmentSection/actions/updateNodeValue.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    tableName: string
+    variableName: string
+    value: NodeValue
+    // TODO support multiple values
+    // values: Array<NodeValue>
+  }
+>(
+  'section/patch/nodeValue',
+  async ({ assessmentName, countryIso, colName, cycleName, sectionName, tableName, variableName, value }) => {
+    const { data } = await axios.patch(
```
Comment (minotogna) [inline client/store/pages/assessmentSection/actions/updateNodeValue.ts:26]: Denounce the Ajax request

File: client/store/pages/assessmentSection/hooks/index.ts [hunk 6]
```diff
@@ -6,3 +6,5 @@ export const useTableSections = (): Array<TableSection> =>
   useAppSelector((state) => state.pages.assessmentSection.tableSections)
 
 export const useTableData = (): TableData => useAppSelector((state) => state.pages.assessmentSection.data)
+
+export const useIsTableDataUpdating = (): boolean => useAppSelector((state) => state.pages.assessmentSection.updating)
```
Comment (minotogna) [inline client/store/pages/assessmentSection/hooks/index.ts:10]: Where is this used ?
Remove it

File: client/store/pages/assessmentSection/slice.ts [hunk 7]
```diff
@@ -28,13 +31,23 @@ export const assessmentSectionSlice = createSlice({
       const countryData = (state.data && state.data[countryIso]) || {}
       state.data = { ...state.data, [countryIso]: { ...payload[countryIso], ...countryData } }
     })
+
+    builder
+      .addCase(updateNodeValue.fulfilled, (state, { payload }) => {
+        state.updating = false
+        state.data = payload
```
Comment (minotogna) [inline client/store/pages/assessmentSection/slice.ts:38]: Remove the fullfilled case

File: client/store/pages/assessmentSection/stateType.ts [hunk 8]
```diff
@@ -4,4 +4,5 @@ import { TableData } from '@meta/data'
 export type AssessmentSectionState = {
   data?: TableData
   tableSections?: Array<TableSection>
+  updating: boolean
```
Comment (minotogna) [inline client/store/pages/assessmentSection/stateType.ts:7]: Remove this

File: client/store/pages/originalDataPoint/actions/updateOriginalDataPoint.ts (hunk 1 above)
Comment (minotogna) [inline client/store/pages/originalDataPoint/actions/updateOriginalDataPoint.ts:11]: Use put. This is an update

File: common/api/endpoint/ApiEndPoint.ts (hunk 2 above)
Comment (minotogna) [reply common/api/endpoint/ApiEndPoint.ts:12]: We discussed this in our meeting a while ago

File: meta/assessment/tableData.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  naturalForestArea?: string
+  naturalForestAreaEstimated?: boolean
+  otherPlantedForestArea?: string
+  otherPlantedForestAreaEstimated?: boolean
+  plantationForestArea?: string
+  plantationForestAreaEstimated?: boolean
+  plantationForestIntroducedArea?: string
+  plantationForestIntroducedAreaEstimated?: boolean
+}
+
```
Comment (minotogna) [inline meta/assessment/tableData.ts:27]: Where are these types used ?

File: server/api/assessment/persistNodeValue.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      countryIso: countryIso as CountryIso,
+      colName,
+      cycle,
+      tableName,
+      user: Requests.getRequestUser(req),
+      variableName,
+      value,
+    })
+    // TODO Remove
+    const tableData = await CycleDataController.getTableData({
```
Comment (minotogna) [inline server/api/assessment/persistNodeValue.ts:28]: Remove this .
Request doesn't return anything, just an empty object

**Review by minotogna — CHANGES_REQUESTED** [review]

File: common/api/endpoint/ApiEndPoint.ts (hunk 2 above)
Comment (rfontanarosa) [reply common/api/endpoint/ApiEndPoint.ts:12]: I don't remember. :(

Comment (minotogna) [reply common/api/endpoint/ApiEndPoint.ts:12]: I see, I misunderstood the question. 
I think the action itself defines the params. would that be enough already?

File: meta/data/tableDatas.ts [hunk 11]
```diff
@@ -1,24 +1,55 @@
...
+  variableName: string
+  colName: string
+  value: NodeValue
+}): TableData => {
+  const { data, countryIso, tableName, variableName, colName, value } = props
+  if (!data?.[countryIso]?.[tableName]?.[variableName]?.[colName]) return data
+  const newData = {
+    ...data,
+  }
+  newData[countryIso][tableName][variableName][colName] = value
```
Comment (minotogna) [inline meta/data/tableDatas.ts:46]: does this work on an initial empty table?

File: server/controller/index.ts [hunk 12]
```diff
@@ -1,4 +1,5 @@
 export { AssessmentController } from './assessment'
+export { CycleDataController } from './cycleData'
+export { SettingsController } from './settings'
 export { UserController } from './user'
 export { UserProviderController } from './userProvider'
```
Comment (minotogna) [inline server/controller/index.ts:5]: @sorja @rfontanarosa should we stop exporting every controller from index? same convention as we agreed with repositories ?

Comment (sorja) [reply server/controller/index.ts:5]: yes

**Review by minotogna — APPROVED** [review]

**Pull request merged**